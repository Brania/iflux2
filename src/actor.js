/**
 * Actor，致敬Erlang，Scala的akka的Actor model
 * Actor, 独立计算的执行单元
 * 我们不共享状态(share state), 只去transform state
 *
 * @flow
 */

//;;;;;;;;flowtype;;;;;;;;;;;;;;;
type Route = {
  [name: string]: Function;
}

type ImmutableState = mixed;


export default class Actor {
  //记录当前的路由信息
  _route: Route;


  /**
   * 定义actor的默认状态
   * @returns {{}}
   */
  defaultState(): Object {
    return {};
  }

  /**
   * actor的mode的receive,被store在dispatch的时候调用
   * @param msg
   * @param state
   * @param param
   * @returns {Object}
   */
  receive(msg: string, state: ImmutableState, param?: any) {
    //this._route是在@Action标记中初始化完成
    const route = this._route || {};
    //获取处理的函数
    const handler = route[msg];

    //如果可以处理返回处理后的结果，否则直接返回state
    return handler ? handler.call(this, state, param) : state;
  }
}
