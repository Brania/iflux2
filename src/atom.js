/**
 * Atom
 *
 * @flow
 */
import {fromJS} from 'immutable';
import Cursor from 'immutable/contrib/cursor';
import {isArray, isStr, isFn} from './util';


/**
 * 原子不可变数据容器
 * @flow
 */
export default class Atom {
  _atom: Object;
  _callbacks: Array<Function>;


  /**
   * 初始化初始的数据结构
   */
  constructor(record: Object) {
    this._callbacks = [];
    this._atom = fromJS(record);
  }


  /**
   * 获取值
   * 1. 如果path为空,就返回所有的值
   * 2. 如果path为字符串或者数组就按照immutable的path返回数据
   * 3. 其他返回空值
   * @param path
   * @returns {*}
   */
  value(path: string|Array<String>) {
    let value = null;
    if(!path) {
      value = this._atom;
    } else if (isStr(path) || isArray(path)) {
      value = this._atom[isStr(path) ? 'get' : 'getIn'](path);
    }
    return value;
  }


  /**
   * 获取cursor
   */
  cursor() {
    return Cursor.from(this._atom, (newState, state, path) => {
      //校验数据是否过期
      if (state != this._atom) {
        console.log && console.log('attempted to alter expired data.');
      }

      //校验有没有数据的变化
      if (newState === state) {
        return;
      }

      this._atom = newState;
      this._callbacks.forEach(cb => cb(newState, path));
    });
  }


  /**
   * 订阅
   */
  subscribe(callback: Function) {
    if (!callback || !isFn(callback)) {
      return;
    }

    //防止重复添加
    for (let i = 0, len = this._callbacks.length; i < len; i++) {
      if (callback === this._callbacks[i]) {
        return;
      }
    }

    this._callbacks.push(callback);
  }


  /**
   * 取消订阅
   */
  unsubscribe(callback: Function) {
    if (!callback || !isFn(callback)) {
      return;
    }

    for (let i = 0, len = this._callbacks.length; i < len; i++) {
      if (callback === this._callbacks[i]) {
        this._callbacks.splice(i, 1);
        break;
      }
    }
  }


  /**
   * 打印出当前的内部数据状态
   */
  pprint() {
    console.log(JSON.stringify(this._atom.toJS(), null, 2));
  }
}
