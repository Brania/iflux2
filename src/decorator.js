/**
 * Action decorator
 *
 * 用于标记Actor中的实力方法，主要的作用是给Actor绑定当前的handler方法
 * 便于Actor的receive方法可以搜索到哪个handler可以处理dispatch过来的事件
 *
 * @param msg 事件名称
 *
 * @flow
 */

'use strict';

export const Action = (msg: string) => (
  target: any,
  props: any,
  descriptor: Object
) => {
  target._route = target._route || {};
  target._route[msg] = descriptor.value;
};
