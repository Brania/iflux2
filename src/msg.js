//@flow

'use strict';

import EventEmitter from 'events';

//全局事件订阅中心
export default new EventEmitter();
