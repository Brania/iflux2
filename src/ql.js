/**
 * 查询语言
 * @flow
 */
import {isQuery, isStr} from './util';


//递增的id
let incrementId = 0;


export class QueryLang {
  _id: number;
  _lang: Array<Object>;
  _name: string|Array<Object>;


  constructor(name: string|Array<Object>, lang: Array<Object>) {
    this._id = ++incrementId;

    //如果第一个参数为字符串，改字符串就代表了该QL的name
    //该name就是为了更好的帮助我们debug调试问题
    if (typeof(name) === 'string' || isStr(name)) {
      this._name = name;
      this._lang = lang;
    } else {
      this._name = '';
      this._lang = name;
    }
  }


  /**
   * 判断当前是不是一个合法的query lang
   * @returns {boolean}
   */
  isValidQuery() {
    return isQuery(this._lang);
  }


  /**
   * 当前的id
   * @returns {number}
   */
  id() {
    return this._id;
  }


  /**
   * 当前的name
   */
  name() {
    return this._name || this._id;
  }


  /**
   * 当前的语法标记
   * @returns {Array.<Object>}
   */
  lang() {
    return this._lang;
  }
}


//Factory Method
export const QL = (name: string|Array<Object>, lang:Array<Object>) => new QueryLang(name, lang);
