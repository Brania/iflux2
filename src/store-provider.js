/**
 * StoreProvider
 * 主要的作用是在Store和React的App之间建立桥梁
 * 将Store初始化,切绑定到React顶层App的上下文
 * @flow
 */
import React from 'react';

type Store = {
  subscribe: (callback: Function) => void;
  unsubscribe: (callback: Function) => void;
};

type Options = {
  debug: boolean;
};

/**
 * WrapperComponent
 * @param AppStore
 * @returns {Function}
 */
export default function connectToStore(
  AppStore: Store,
  opts: Options = {}
) {
  return function (Component: ReactClass<{}>) {
    return class StoreContainer extends React.Component {
      //关联的store
      _store: Object;
      //当前的组件状态
      _isMounted: boolean;

      static childContextTypes = {
        store: React.PropTypes.object
      };

      getChildContext: Function = (): Object => {
        return {
          store: this._store
        };
      };

      constructor(props: Object) {
        super(props);
        //如果是debug状态
        if (opts.debug) {
          console.group(`StoreProvider(${Component.name}) in debug mode.`);
        }

        //初始化当前的组件状态
        this._isMounted = false;
        //初始化Store
        this._store = new AppStore(opts);
      }

      componentWillMount() {
        this._isMounted = false;
        if (opts.debug) {
          console.log(`${Component.name} will mount...`);
          console.time('first-render-time');
        }
      }

      componentDidMount() {
        if (opts.debug) {
          console.log(`${Component.name} did mount`);
          console.timeEnd('first-render-time');
          console.groupEnd();
        }
        this._isMounted = true;
        this._store.subscribe(this._handleStoreChange);
      }

      componentWillUpdate() {
        this._isMounted = false;
        if (opts.debug) {
          console.group(`${Component.name} will update`);
          console.time('update-render-time');
        }
      }

      componentDidUpdate() {
        this._isMounted = true;
        if (opts.debug) {
          console.log(`${Component.name} did update`)
          console.timeEnd('update-render-time')
          console.groupEnd();
        }
      }

      componentWillUnmount() {
        this._store.unsubscribe(this._handleStoreChange);
      }


      render() {
        return (
          <Component {...this.props} store={this._store} />
        );
      }


      _handleStoreChange = () => {
        if (this._isMounted) {
          this.forceUpdate();
        }
      };
    }
  }
}
