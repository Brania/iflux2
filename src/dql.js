import {QL, QueryLang} from './ql';
import {isArray, isStr, isQuery} from './util';


/**
 * 动态的QueryLang
 * @flow
 */
export class DynamicQueryLang {
  _ctx: Object;
  _ql: Object;


  constructor(name: string | Array<Object>, lang: Array<Object>) {
    this._ctx = {};
    this._ql = QL(name, lang);
  }


  /**
   * 分析路径中的动态元素，然后根据上下文替换
   * 
   * @param  {Array<Object>} lang
   */
  analyserLang(ql: Object) {
    //校验query的合法性
    if (!ql.isValidQuery()) {
      throw new Error(`PQL invalid partial query-lang`);
    }

    //获取语法结构
    let lang = ql.lang();
    for (let i = 0, len = lang.length - 1; i < len; i++) {
      //获取当前的路径
      let path = lang[i];
      if (isStr(path) && path[0] === '$') {
        //重新赋值
        lang[i] = this._ctx[path.substring(1)];
      } else if (isArray(path)) {
        for (let j = 0, len = path.length; j < len; j++) {
          let path = lang[i][j];
          if (isStr(path) && path[0] === '$') {
            //重新赋值
            lang[i][j] = this._ctx[path.substring(1)];
          }
        }
      } else if (path instanceof DynamicQueryLang) {
        //递归一次
        this.analyserLang(path._ql);
        lang[i] = path._ql;
      }
    }
  }


  /**
   * 返回可用的query lang
   */
  ql() {
    this.analyserLang(this._ql);
    return this._ql;
  }


  /**
   * 设置上下文
   * @param  {Object} ctx
   */
  context(ctx: Object) {
    this._ctx = ctx;
    return this;
  }
}


/**
 * 工厂函数
 * @param  {string|Array<Object>} name
 * @param  {Array<Object>} lang
 */
export const DQL = (name: string | Array<Object>, lang: Array<Object>) => new DynamicQueryLang(name, lang);