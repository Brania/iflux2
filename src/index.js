import msg from './msg';
import Actor from './actor';
import StoreProvider from './store-provider';
import Relax from './relax';
import Store from './store';
import {Action, CtxStoreName} from './decorator';
import Atom from './atom';
import {QL} from './ql';
import {DQL} from './dql';
import Validator from './validator';
import Util from './util';

export {
  QL,
  DQL,
  msg,
  Atom,
  Action,
  CtxStoreName,
  Actor,
  Util,
  Relax,
  Store,
  StoreProvider,
  Validator
};
