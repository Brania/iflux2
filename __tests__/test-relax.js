import React from 'react'
import renderer from 'react-test-renderer';
import Relax from '../src/relax'
import StoreProvider from '../src/store-provider';
import Store from '../src/store'
import Actor from '../src/actor'
import {Action} from '../src/decorator'
jest.mock('react-dom');

//;;;;;;;;;;Actor;;;;;;;;;;;;;;;;;;
class TestActor extends Actor {
  defaultState() {
    return {
      text: 'hello world'
    }
  }

  @Action('init')
  init(state, text) {
    return state.set('text', text);
  }
}


//;;;;;;;;;;;;;;;Store;;;;;;;;;;;;;;;
class AppStore extends Store  {
  bindActor() {
    return [
      new TestActor()
    ]
  }

  init = () => {
    this.dispatch('init', 'hello iflux2');
  };
}

//;;;;;;;;;;;;;;;;Relax;;;;;;;;;;;;;;;
@Relax
class Text extends React.Component {
  static defaultProps = {
    text: ''
  };

  render() {
    return (
      <div>{this.props.text}</div>
    )
  }
}

//;;;;;;;;;;;;;;;;;;root;;;;;;;;;;;;;
@StoreProvider(AppStore)
class HelloApp extends React.Component {
  componentWillMount() {
    this.props.store.init();
  }

  render() {
    return (
      <Text/>
    )
  }
}

//;;;;;;;;;;;test;;;;;;;;;;;;;;;;;;;;;
test('test store provider and relax', () => {
  const tree = renderer.create(<HelloApp/>).toJSON();
  expect(tree).toMatchSnapshot();
});