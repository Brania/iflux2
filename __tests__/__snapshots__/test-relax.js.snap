exports[`test test store provider and relax 1`] = `
<div>
  hello iflux2
</div>
`;
